### ISSUE TITLE
Prevent user data to be written to database when POST Register User Failed 

### ISSUE THREAD

**Issue by mtreacy002** [anitab-org/mentorship-backend#461]
**Describe the bug**
Create user using POST Register adds data to database when the HTTP response is not 200. This is found when local database is used.

**To Reproduce**
1. Make sure you set FLASK_ENVIRONMENT_CONFIG to local in ".env".
2. Run python run.py from the terminal.
3. Go to the link created by step above.
4. Go to 'POST/register' and add user (that is not already in the database and using wrong pattern).
5. You should receive response 500.
6. Check the database by opening new terminal and run 'sqlite3 local_data.db;'
7. On sqlite3 shell run 'select * from users;'
8. You should see that the user has been added to the record regardless of the process being failed.

**Expected behavior**
The user should not get added to the database as it did not passed validation.

**Screenshots**

- Swagger API with local database - error on POST/Register.

![Screen Shot 2020-02-27 at 11 11 36 am](https://user-images.githubusercontent.com/29667122/75400160-929a5680-5952-11ea-99f2-663b940ae69c.png)

- Sqlite local data added user regardless failed process.
<img width="646" alt="Screen Shot 2020-02-27 at 11 12 57 am" src="https://user-images.githubusercontent.com/29667122/75400256-d55c2e80-5952-11ea-8cde-47acd484714d.png">

**Desktop (please complete the following information):**
 - OS: Mac OS 
 - Browser: Safari
 - Version (is this referring to Mac OS version??): Mojave 10.14.6 

**Additional context**
Might need some validation prior to allowing user to be added to database that will cancel the process if HTTP returned with other than 200.

**Comment by isabelcosta** [comment]
@mtreacy002 I knew this issue already but did not have time to report the issue.
So thank you so so much for reporting this bug, following our template to the fullest and providing the detailed context of the problem! 🎉 🙌  Very good issue description 👏

**Comment by mtreacy002** [comment]
Thanks for the comment, @isabelcosta. I'm glad I could give a bit of support.

PS: When I tried to Login using the wrong credentials that created from the process above, it responded with HTTP 403 as email is not yet confirm. This email I believe will never be sent to the user as the registration process is failed.
![Screen Shot 2020-02-27 at 11 41 51 am](https://user-images.githubusercontent.com/29667122/75401557-95974600-5956-11ea-8ba1-e074f136cac3.png)

**Comment by mtreacy002** [comment]
@isabelcosta, Is it ok if I try to work on this issue?

**Comment by isabelcosta** [comment]
@mtreacy002 go ahead :D thank you so much for contributing 🙌

**Comment by mtreacy002** [comment]
@isabelcosta, can I confirm that currently we don't have password validation stated anywhere in the project? I have checked 'app/utils/validation_utils.py' and can't see one that check password pattern/regex. It seems like we only consider the length of the password (between 8-25). Is it intended or should I add some regex validation for password. If we choose the later, what would the requirement be? Can you please advise? Thanks beforehand.

**Comment by mtreacy002** [comment]
@isabelcosta, after debugging the code I found out that the culprit is not on the data creation itself, but when sending email confirmation. 

<img width="713" alt="Screen Shot 2020-02-28 at 1 13 50 am" src="https://user-images.githubusercontent.com/29667122/75453213-aaf88880-59c7-11ea-99be-5ff9a9206d59.png">

The error was caused by the Gmail '**Less secure app access**' was set to 'OFF' (I had already set the **2 steps verification** to 'OFF'). When I allowed the access from less secure app, the error disappeared and I received the email confirmation.

I'm now thinking that we might not need to do anything to the existing code (except for adding password validation, which is a separate issue). We just need to inform the user (on the UI) about the required settings for the  2 steps verification and Less secure app access above, until we can find ways to work around those settings. 

The other option is to delete the new created user from the database if the 'send-email' return with bad response and show the error message to user that verification email failed to send. 

Can you please advise which option we should take? Thanks

**Comment by mtreacy002** [comment]
I was curious and looked at Front-End. It seems like this issue also related to [issue#105](https://github.com/anitab-org/mentorship-android/issues/105) on Front-End. It will be helpful if we could send error message to user to inform them why they haven't received the email (which 2 settings above being the potential issue).

**Comment by isabelcosta** [comment]
> The error was caused by the Gmail 'Less secure app access' was set to 'OFF' (I had already set the 2 steps verification to 'OFF'). When I allowed the access from less secure app, the error disappeared and I received the email confirmation.

I understand this as an issue (documented on #233), but this is on the developer's gmail account settings. So the error being on the server's part, the user should not be created if the "email sending" functionality is failing. So I think we do have to avoid that the user creation if we detect, email failed to be sent.

> I'm now thinking that we might not need to do anything to the existing code (except for adding password validation, which is a separate issue). We just need to inform the user (on the UI) about the required settings for the 2 steps verification and Less secure app access above, until we can find ways to work around those settings.

About this I think we should not return a error message mentioning the setup of email, but just mention something useful from the user of Mentorship System perspective. Something like "we were unable to complete the request, try again soon", and if we want to be more clear about the setup, we can document this problem for developers to be aware. 🤔

> The other option is to delete the new created user from the database if the 'send-email' return with bad response and show the error message to user that verification email failed to send.

This seems like a good option :)

> Can you please advise which option we should take? Thanks

So you're last option, seems the good option, and for that you would have to change the code somehow to detect, when something fails from the email side, let the user know about it and avoid the user creation in the database. Please let me know what you think about it @mtreacy002 :)

**Comment by isabelcosta** [comment]
> I was curious and looked at Front-End. It seems like this issue also related to [issue#105](https://github.com/anitab-org/mentorship-android/issues/105) on Front-End. It will be helpful if we could send error message to user to inform them why they haven't received the email (which 2 settings above being the potential issue).

What do you think should be done?

**Comment by mtreacy002** [comment]
Yep, I agree that preventing user from being recorded is the way to go. I think we might have to swift the data validation fields to the front-end so only correct data will be process by the backend (maybe like disabling Register button until the correct data are inserted). Once the data is correct, user is allowed to click Register (Register button enabled). This also helps reduce server workload and make our app much faster (maybe??).

On the back-end, the first thing we'll do probably sending the test email saying 'Thank you for registering, you will receive confirmation email soon' just to test the email is accessible. When this is failed we can show error message to user saying "we were unable to complete the request, try again soon" as your post above. 

Only after test email passed we will save user data in database then send verification email. This way we will prevent unnecessary process of entering data, deleting it if verification email fail to send and re-enter the data again on the settings are right. 

This is just my thought. What is your thoughts on this, @isabelcosta?

**Comment by mtreacy002** [comment]
On second thought, we might not need to test send email at all, but just catch the error received as the result of sending verification email and show error message as you said. If this is pass, then we record the data. 

Sorry for all the long posts above. Just thinking outloud here, 😄. I'll start doing the work now... 😊

**Comment by mtreacy002** [comment]
Done. The server now won't record data when verification email failed to send.
An error message also will be displayed to user. Will push this to repo soon.

<img width="567" alt="Screen Shot 2020-02-28 at 11 59 13 am" src="https://user-images.githubusercontent.com/29667122/75500671-c6907d00-5a21-11ea-9462-75acb94e4b06.png">

**Comment by robotjellyzone** [comment]
Hi @mtreacy002 @isabelcosta as said in the PR of this issue also that i am unable to reproduce this error , i don't know why you are getting this error, as i have registered the user successfully !! .As can be seen from below screenshots : 

Using swagger to register :-

![2020-04-13](https://user-images.githubusercontent.com/36916536/79077571-53686f00-7d20-11ea-9b48-c7437c8d8799.png)

Using sqlite3 to show that the user data is stored and registered :-
![2020-04-13 (2)](https://user-images.githubusercontent.com/36916536/79077591-7b57d280-7d20-11ea-819e-2a04378b3007.png)

Using swagger to show the user is confirmed by clicking link sent on temporary mail :-
![2020-04-13 (1)](https://user-images.githubusercontent.com/36916536/79077623-9de9eb80-7d20-11ea-80fc-eba5462df084.png)
